src/util/emitUtils.js

```diff
@@ -578,22 +580,29 @@ function emitToClients(io, nsp, rooms, key, obj) {
  */
 function doEmit(nsp, key, obj) {
   const newObjectAsString = getNewObjAsString(key, obj); // { key: {new: obj }}
-  const numClients = Object.values(nsp.connected).length;
-  if (key.startsWith('refocus.internal.realtime.sample')) {
-    tracker.trackEmit(obj.name, obj.updatedAt, numClients);
-  }
   pubSubStats.trackEmit(key, obj);
-  if (!toggle.isFeatureEnabled('enableClientStats')) { // dont enter the else if its not a sample
-    nsp.emit(key, newObjectAsString);
-  } else if (key.startsWith('refocus.internal.realtime.sample')) {
+  if (toggle.isFeatureEnabled('enableClientStats') && key.startsWith('refocus.internal.realtime.sample')) { // dont enter the else if its not a sample
+    let numClients = 0;
+    const rooms = nsp.rooms.slice(0);
+    nsp.rooms = [];
```

Comment:
Why are you setting rooms to empty?

Reply:
`nsp.rooms` is an array of the id's of the rooms that the namespace will emit the next event to, when we call `nsp.emit()`.

When `nsp.to(room)` is called in the loop on line 565 the room id is pushed to `nsp.rooms` and then when `nsp.emit()` is called the message is emitted to only the rooms in this array and then the array is set to [] again. (When the array is [ ] then `nsp.emit()` sends the event to all connected clients.)

Since we're not calling `nsp.emit()` and looping through the sockets in an unconventional way this array never gets cleared, we need to manually clear the rooms array or this will have an effect the next time `nsp.emit()` is called.

Reply:
Ooooh, that's interesting. So basically we have two completely different approaches to doing filtering, written at two different times, that were combined together without fully thinking through how they would interact, which results in unexpected issues and requires a hack like this to fix.

Makes me think the two approaches should be completely separate: one function for normal emits, one function for emits with callbacks, and emitClients calls one or the other based on the toggle. That way we can set the rooms only when we know they're going to be used for the emit.

Reply:
Yeah, well the main issue is that until now the code to emit with callbacks wasn't filtering at all since it was just looping through `nsp.connected`; in this change we're using this hacky method of comparing every socket in `nsp.connected` to see if they're in any of the `nsp.rooms` and if we are then we emit them. So we're using rooms in both scenarios really and I don't think there's any way around it if we want to attach callbacks to each emit like this.
I would be agree about separate methods since this function is getting quite long.

Reply:
Well, there's no way around using the room *names* in both scenarios, but that's fine because we already have them anyway, from being setup in emitter.js. But we don't have to use the room *objects* for the callback emit - we could just iterate over the list of room names that we already have. That way we wouldn't have to set and clear them on the namespace object.

Reply:
Oh, but `nsp.rooms ` is just an array of the room names anyway, we need to set them for all the non sample events anyway so it's pretty much equivalent to passing in the array from emitter.js

Reply:
Ok right, maybe calling it a room object isn't accurate, but what I mean is setting the room on the namespace object; in other words, changing the state of the socket.io server. My point is that it's not a good idea to change the state and then use the hack to change it back, it would be better to not change that state at all when we don't need to. And if we had two separate functions, we could just pass in the room names to the callback emit without having to change the state in that case.

Reply:
Added function `doEmitWithTracking` with changes you recommended